### PR TITLE
feat(cmcd): dedup state-change events + auto-trigger from update()

### PR DESCRIPTION
**Pull request by littlespex**
## Summary

- Suppresses consecutive `ps`/`pr`/`c`/`b`/`bc` events with unchanged values (matches CTA-5004-B intent for these as state-transition events).
- `CmcdReporter.update()` now auto-fires the corresponding state-change event when a tracked field (`sta`, `pr`, `cid`, `bg`, `br`) changes — collapses the documented two-step `update()` + `recordEvent()` pattern into a single call.
- `CmcdReporter.recordEvent(<state-event>, { field })` now write-throughs the dedup field into `this.data`, unifying persistence with `update()`.
- Adds `CmcdEventType.PLAYBACK_RATE` (token `'pr'`) — defined in CTA-5004-B and referenced in `CmcdEvent.ts` docs but missing from the type.

## Architecture

- Declarative `STATE_FIELDS` dispatch table co-locates field name, event type, and equality function. Adding a future dedup-able field is a one-row change.
- `STATE_FIELDS_BY_EVENT` Map gives O(1) lookup in `recordEvent` (consistent with `new Set(...)` / `new Map(...)` patterns elsewhere in the package).
- `lastEmitted` private field tracks the wire-emission baseline — separate from `this.data` because `update()` can mutate persistent data silently. Cleared on `sid` change in `resetSession`.
- Module-private `cmcdObjectTypeListEqual` helper handles `br`'s heterogeneous list (numbers + `SfItem<number, ExclusiveRecord<CmcdObjectType, boolean>>`). Order-sensitive; spurious-emit is the safer failure mode versus missed-emit.

## Backward compatibility

Strictly additive. The documented two-step pattern (`update({ sta }); recordEvent(PLAY_STATE)`) still works — the second call is harmlessly deduped against the auto-trigger from the first.

## Spec & plan

- Spec: `docs/superpowers/specs/2026-05-22-cmcd-state-event-dedup-design.md`
- Plan: `plans/cmcd-state-event-dedup/plan.md`

## Test plan

- [ ] CI green: `npm run build -w libs/cmcd && npm run typecheck && npm run lint && npm test -w libs/cmcd`
- [ ] Verify 263 tests (261 pass / 2 todo / 0 fail), with 22 new tests in the `state-change dedup` describe block
- [ ] Read the updated `libs/cmcd/docs/user-guide.md` auto-trigger example
- [ ] Verify CHANGELOG entry under [2.4.0]
- [ ] Smoke-check existing tests that use `update() + recordEvent()` still pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: libs/cmcd/src/CmcdReporter.ts [hunk 1]
```diff
@@ -209,17 +292,57 @@ export class CmcdReporter {
 		// msd is tracked separately via this.msd and sent once per target,
 		// so it is stripped from the persistent data store after each update.
 		this.data = { ...this.data, ...data, msd: undefined }
+
+		// Auto-trigger state-change events for any tracked field whose value
+		// differs from the last wire-emitted value. Comparing against lastEmitted
+		// (not the pre-merge value) ensures correctness after a session reset,
+		// and unifies the comparison basis with recordEvent's internal dedup.
+		for (const entry of STATE_FIELDS) {
+			if (entry.field in data && !entry.equal(this.data[entry.field], this.lastEmitted[entry.field])) {
```
[the remark this answers is not in the document]
Comment (littlespex) [reply libs/cmcd/src/CmcdReporter.ts:312]: Confirmed and fixed in 3a4df6a1f. Added a guard in `recordEvent` that drops emission whenever the effective dedup field is `undefined` (after write-through), regardless of `lastEmitted` state. This covers both "never set" and "explicitly cleared via update({field:undefined})" paths. Added a regression test (`drops PLAY_STATE when sta is explicitly cleared to undefined`) and updated the spec edge-case section to call out both scenarios.

File: libs/cmcd/src/CmcdReporter.ts [hunk 2]
```diff
@@ -193,7 +264,19 @@ export class CmcdReporter {
...
+	 * (`sta`, `pr`, `cid`, `bg`, `br`), if the new value differs from the
+	 * current value, the corresponding state-change event is automatically
+	 * fired (subject to dedup against the last-emitted value).
+	 *
+	 * Multi-field updates fire multiple events in the order: `sta` → `pr` →
+	 * `cid` → `bg` → `br`. The order of keys in the input object does not
+	 * affect the firing order.
+	 *
+	 * A `sid` change resets the dedup baseline so the next state-change
+	 * event in the new session always emits.
```
[the remark this answers is not in the document]
Comment (littlespex) [reply libs/cmcd/src/CmcdReporter.ts]: Fixed in 3a4df6a1f. The TSDoc now describes the comparison basis as the last-reported value (matching the implementation), with a sentence explaining why this differs from the previous persisted value (sid-reset correctness). The implementation was correct; the doc was stale from an earlier draft.

File: libs/cmcd/docs/user-guide.md [hunk 3]
```diff
@@ -190,11 +190,27 @@ Use the `recordEvent()` method to record player events for event-mode reporting.
...
+reporter.update({ br: [5000] });
+// → fires CmcdEventType.BITRATE_CHANGE
+
+// Consecutive same-value updates are deduplicated.
+reporter.update({ sta: "a" }); // emits (sta changed from "p" to "a")
+reporter.update({ sta: "a" }); // dropped (unchanged)
+
+// recordEvent() still works for attaching extra context at a transition,
+// and for non-state events like CUSTOM_EVENT and ERROR.
+reporter.recordEvent(CmcdEventType.PLAY_STATE, { sta: "a", bl: [3000] });
```
[the remark this answers is not in the document]
Comment (littlespex) [reply libs/cmcd/docs/user-guide.md]: Fixed in 3a4df6a1f. The example now shows a same-value `update` being deduplicated (without transitioning to a new value), then `recordEvent(PLAY_STATE, { sta: "a", bl: [3000] })` introducing the transition with extra context. Trace: initial `update({sta:"p"})` sets lastEmitted.sta="p"; the followup `update({sta:"p"})` is dropped; `recordEvent` write-throughs sta="a" and the p→a transition emits with bl included in the payload.

File: libs/cmcd/src/CmcdReporter.ts [hunk 4]
```diff
@@ -47,6 +48,75 @@ function createEncodingOptions(reportingMode: CmcdReportingMode, config: CmcdRep
...
+function cmcdObjectTypeListEqual(a: CmcdObjectTypeList, b: CmcdObjectTypeList): boolean {
+	if (a === b) return true
+	if (a.length !== b.length) return false
+
+	for (let i = 0; i < a.length; i++) {
+		const ai = a[i]
+		const bi = b[i]
+		if (ai === bi) continue
+		if (typeof ai === 'number' || typeof bi === 'number') return false
+
```
[the remark this answers is not in the document]
Comment (littlespex) [reply libs/cmcd/src/CmcdReporter.ts:88]: Confirmed and fixed in db17911cc.

Root cause: after `recordEvent` ran `Object.assign(this.lastEmitted, { br: current })`, `lastEmitted.br === this.data.br === caller's array`. Any in-place mutation of the caller's array poisoned the dedup baseline; the `a === b` shortcut in `cmcdObjectTypeListEqual` then masked the change as a missed emit (removing the shortcut alone wouldn't have helped, since element-wise comparison of an array against itself still reports equal).

Fix: added a `snapshot` function to each `STATE_FIELDS` entry (identity for scalars, `.slice()` for `br`) and use it when recording into `lastEmitted`. The dedup baseline is now an independent array copy, so caller mutation has no effect on it. Added a regression test (`emits when caller mutates the br array in place between updates`) covering the push/mutate-then-update path; all 263 tests pass (0 fail, 2 todo).

File: libs/cmcd/src/CmcdReporter.ts [hunk 5]
```diff
@@ -209,17 +295,66 @@ export class CmcdReporter {
...
+			if (current === undefined) {
+				return
+			}
+
+			if (entry.equal(current, this.lastEmitted[field])) {
+				return
+			}
+
+			Object.assign(this.lastEmitted, { [field]: current })
+		}
```
[the remark this answers is not in the document]
Comment (littlespex) [reply libs/cmcd/src/CmcdReporter.ts:364]: The technical claim is correct — `prepareCmcdData` enforces `e`/`ts`/`cen`/`v` post-filter but not the dedup field — but I'd like to keep this out of scope for this PR.

1. **Pre-existing behavior, not a regression.** The per-target `enabledKeys` filter has always allowed operators to drop any key. Before this PR, `recordEvent(PLAY_STATE, { sta: 'p' })` against a target whose `enabledKeys` omits `sta` would produce the same malformed `e=ps` without `sta`. This PR's guard tightens the "value undefined everywhere" case (a programming error in the caller); it doesn't claim to enforce per-target spec compliance.

2. **Per-target filter is operator intent.** `enabledKeys` is the configured contract for "do not emit this key to this target." If an operator enables PLAY_STATE events but omits `sta` from `enabledKeys`, they've explicitly opted out of `sta` — possibly deliberately (PII, payload size). The library has no clean way to distinguish misconfiguration from deliberate redaction.

3. **The right fix lives in `prepareCmcdData`.** Your suggestion (a) — force-include the dedup field for state-change events, symmetric with how `e`/`ts`/`cen` are enforced today — is the cleanest path and touches the encoder, not dedup. Suggestion (b) (per-target drop) would conflict with the operator-intent reading above.

I'd prefer to file this as a follow-up so the encoder change gets its own PR + test matrix (each of the 5 state-change events × enabled/disabled dedup key × event/request mode). Will open an issue referencing this thread if (a) sounds right.

[the remark this answers is not in the document]
Comment (littlespex) [reply libs/cmcd/src/CmcdReporter.ts:364]: Filed as #368 for the encoder-side fix. Resolving here since the dedup work in this PR is unchanged.
